pkg/server/server.go

```diff
@@ -0,0 +1,159 @@
+	return &Server{
+		log:         log.WithField("module", "certificate_provider"),
+		client:      kubeOptions.CMClient,
+		auther:      kubeOptions.Auther,
+		issuerRef:   cmOptions.IssuerRef,
+		preserveCRs: cmOptions.PreserveCRs,
+	}
+}
+
+// Run is a blocking func that will run the client facing certificate service
+func (s *Server) Run(ctx context.Context, tlsConfig *tls.Config, listenAddress string) error {
+	// Setup the grpc server using the passed TLS config
+	creds := credentials.NewTLS(tlsConfig)
+	grpcServer := grpc.NewServer(grpc.Creds(creds))
+
+	// listen on the configured address
+	listener, err := net.Listen("tcp", listenAddress)
+	if err != nil {
+		return fmt.Errorf("failed to listen %s: %v", listenAddress, err)
+	}
+
+	// register certificate service grpc API
+	securityapi.RegisterIstioCertificateServiceServer(grpcServer, s)
+
+	// handle termination gracefully
+	go func() {
+		<-ctx.Done()
+		s.log.Info("shutting down grpc server")
+		grpcServer.GracefulStop()
+		s.log.Info("grpc server stopped")
+	}()
+
+	s.log.Infof("grpc serving on %s", listener.Addr())
+
+	return grpcServer.Serve(listener)
+}
+
+// CreateCertificate is the istio grpc API func, to authenticate, authorize,
+// and sign CSRs requests from istio clients.
+func (s *Server) CreateCertificate(ctx context.Context, icr *securityapi.IstioCertificateRequest) (*securityapi.IstioCertificateResponse, error) {
+	// authn incoming requests, and build concatenated identities for labelling
+	identities, ok := s.authRequest(ctx, []byte(icr.Csr))
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "request authenticate failure")
+	}
+
+	// Build cert-manager CertificateRequest based on the configured issuer
+	cr := &cmapi.CertificateRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			// Random non-conflicted name
+			GenerateName: "istio-",
+			Annotations: map[string]string{
+				// Label identities to resource for auditing
+				IdentitiesAnnotationKey: identities,
+			},
+		},
+		Spec: cmapi.CertificateRequestSpec{
+			Duration: &metav1.Duration{
+				// Add during which was requested from the client.
+				// TODO: We should have a configurable maximum that the duration can
+				// be. Take smaller of the two.
+				Duration: time.Duration(icr.ValidityDuration) * time.Second,
+			},
+			IsCA:      false,
+			Request:   []byte(icr.Csr),
+			Usages:    []cmapi.KeyUsage{cmapi.UsageClientAuth, cmapi.UsageServerAuth},
+			IssuerRef: s.issuerRef,
+		},
+	}
+
+	// Create CertificateRequest and wait for it to become ready
+	cr, err := s.client.Create(ctx, cr, metav1.CreateOptions{})
+	if err != nil {
+		s.log.Errorf("failed to create CertificateRequest for %q: %s",
+			identities, err)
+		return nil, status.Error(codes.Internal, "failed to sign certificate request")
+	}
+
+	log := util.LogWithCertificateRequest(s.log, cr)
+
+	cr, err = util.WaitForCertificateRequestReady(ctx, log, s.client, cr.Name, time.Second*30)
```

Comment:
Should this be configurable? Is there some problem if certificaterequests take longer to be signed? I guess that istio will timeout waiting for us at some point anyway?

Reply:
Yes, I think we should extend this though anyhow. As you say, we will start to balloon CRs if the issuer is broken. I will increase the time limit and delete if errored?

Reply:
That seems like a fine start. There's a flag to retain the certificaterequests for debugging.